File: gui/widgets/file_select_page.py
"""
Page 1: File Import (dev doc §12.1)

Controls: ESI path, SDO path, template path, output dir
Buttons: Browse, Parse
After parse: shows device identity summary
"""
from pathlib import Path
from PyQt5.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QGridLayout,
    QLabel, QLineEdit, QPushButton, QFileDialog,
    QGroupBox, QTextEdit, QSizePolicy,
)
from PyQt5.QtCore import pyqtSignal


class FileSelectPage(QWidget):
    """File selection and initial parsing page."""

    # Emitted when user clicks "Parse" — carries the file paths
    parse_requested = pyqtSignal(dict)
    # Emitted after successful parse — carries parsed info dict
    parse_done = pyqtSignal(dict)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._init_ui()

    def _init_ui(self):
        layout = QVBoxLayout(self)
        layout.setSpacing(12)

        # ── File selection group ───────────────
        file_group = QGroupBox("Input Files")
        grid = QGridLayout(file_group)

        self._esi_edit = self._make_file_row(grid, 0, "ESI.XML:", "XML files (*.xml);;All (*)")
        self._sdo_edit = self._make_file_row(grid, 1, "SDO.XML:", "XML/TXT files (*.xml *.txt);;All (*)")
        self._tpl_edit = self._make_file_row(grid, 2, "Template xlsx:", "Excel files (*.xlsx);;All (*)")
        self._out_edit = self._make_dir_row(grid, 3, "Output dir:")

        layout.addWidget(file_group)

        # ── Parse button ───────────────────────
        btn_row = QHBoxLayout()
        btn_row.addStretch()
        self._parse_btn = QPushButton("Parse")
        self._parse_btn.setMinimumWidth(120)
        self._parse_btn.clicked.connect(self._on_parse)
        btn_row.addWidget(self._parse_btn)
        layout.addLayout(btn_row)

        # ── Parse result display ───────────────
        result_group = QGroupBox("Parse Result")
        result_layout = QVBoxLayout(result_group)
        self._result_text = QTextEdit()
        self._result_text.setReadOnly(True)
        self._result_text.setMaximumHeight(180)
        result_layout.addWidget(self._result_text)
        layout.addWidget(result_group)

        layout.addStretch()

    # ── UI helpers ─────────────────────────────

    def _make_file_row(self, grid, row, label, filter_str):
        """Add a label + line edit + browse button row."""
        lbl = QLabel(label)
        edit = QLineEdit()
        btn = QPushButton("Browse...")
        btn.setMaximumWidth(80)
        btn.clicked.connect(lambda: self._browse_file(edit, filter_str))
        grid.addWidget(lbl, row, 0)
        grid.addWidget(edit, row, 1)
        grid.addWidget(btn, row, 2)
        return edit

    def _make_dir_row(self, grid, row, label):
        """Add a label + line edit + browse-dir button row."""
        lbl = QLabel(label)
        edit = QLineEdit("out")
        btn = QPushButton("Browse...")
        btn.setMaximumWidth(80)
        btn.clicked.connect(lambda: self._browse_dir(edit))
        grid.addWidget(lbl, row, 0)
        grid.addWidget(edit, row, 1)
        grid.addWidget(btn, row, 2)
        return edit

    def _browse_file(self, edit, filter_str):
        path, _ = QFileDialog.getOpenFileName(self, "Select file", "", filter_str)
        if path:
            edit.setText(path)

    def _browse_dir(self, edit):
        path = QFileDialog.getExistingDirectory(self, "Select output directory")
        if path:
            edit.setText(path)

    # ── Actions ────────────────────────────────

    def _on_parse(self):
        """Collect file paths and emit parse_requested signal."""
        paths = {
            "esi_path": self._esi_edit.text().strip() or None,
            "sdo_path": self._sdo_edit.text().strip() or None,
            "template_path": self._tpl_edit.text().strip() or None,
            "output_dir": self._out_edit.text().strip() or "out",
        }
        if not paths["esi_path"] and not paths["sdo_path"]:
            self._result_text.setPlainText("Error: Provide at least one of ESI or SDO file.")
            return
        self.parse_requested.emit(paths)

    def show_parse_result(self, info: dict):
        """Display parse result summary (called from main_window)."""
        lines = []
        for key, val in info.items():
            lines.append(f"{key}: {val}")
        self._result_text.setPlainText("\n".join(lines))

    def get_paths(self) -> dict:
        """Return current file paths."""
        return {
            "esi_path": self._esi_edit.text().strip() or None,
            "sdo_path": self._sdo_edit.text().strip() or None,
            "template_path": self._tpl_edit.text().strip() or None,
            "output_dir": self._out_edit.text().strip() or "out",
        }

File: gui/src/pages/fileselectpage.cpp
/**
 * 页1: 文件导入 — 实现。
 */
#include "fileselectpage.h"
#include <QVBoxLayout>
#include <QGridLayout>
#include <QGroupBox>
#include <QLabel>
#include <QFileDialog>

FileSelectPage::FileSelectPage(QWidget *parent) : QWidget(parent) {
    auto *layout = new QVBoxLayout(this);
    layout->setSpacing(12);

    // ── 文件选择组 ──────────────────────────
    auto *fileGroup = new QGroupBox("输入文件");
    auto *grid = new QGridLayout(fileGroup);

    // ESI 行
    grid->addWidget(new QLabel("ESI.XML:"), 0, 0);
    m_esiEdit = new QLineEdit;
    grid->addWidget(m_esiEdit, 0, 1);
    auto *esiBtn = new QPushButton("浏览...");
    esiBtn->setMaximumWidth(80);
    connect(esiBtn, &QPushButton::clicked, this, [this]() {
        browseFile(m_esiEdit, "XML files (*.xml);;All (*)");
    });
    grid->addWidget(esiBtn, 0, 2);

    // SDO 行
    grid->addWidget(new QLabel("SDO 文件:"), 1, 0);
    m_sdoEdit = new QLineEdit;
    grid->addWidget(m_sdoEdit, 1, 1);
    auto *sdoBtn = new QPushButton("浏览...");
    sdoBtn->setMaximumWidth(80);
    connect(sdoBtn, &QPushButton::clicked, this, [this]() {
        browseFile(m_sdoEdit, "XML/TXT files (*.xml *.txt);;All (*)");
    });
    grid->addWidget(sdoBtn, 1, 2);

    // 模板行
    grid->addWidget(new QLabel("模板 xlsx:"), 2, 0);
    m_tplEdit = new QLineEdit;
    grid->addWidget(m_tplEdit, 2, 1);
    auto *tplBtn = new QPushButton("浏览...");
    tplBtn->setMaximumWidth(80);
    connect(tplBtn, &QPushButton::clicked, this, [this]() {
        browseFile(m_tplEdit, "Excel files (*.xlsx);;All (*)");
    });
    grid->addWidget(tplBtn, 2, 2);

    // 输出目录行
    grid->addWidget(new QLabel("输出目录:"), 3, 0);
    m_outEdit = new QLineEdit("out");
    grid->addWidget(m_outEdit, 3, 1);
    auto *outBtn = new QPushButton("浏览...");
    outBtn->setMaximumWidth(80);
    connect(outBtn, &QPushButton::clicked, this, [this]() {
        browseDir(m_outEdit);
    });
    grid->addWidget(outBtn, 3, 2);

    layout->addWidget(fileGroup);

    // ── 解析按钮 ────────────────────────────
    auto *btnRow = new QHBoxLayout;
    btnRow->addStretch();
    auto *parseBtn = new QPushButton("解析");
    parseBtn->setMinimumWidth(120);
    connect(parseBtn, &QPushButton::clicked, this, &FileSelectPage::onParseClicked);
    btnRow->addWidget(parseBtn);
    layout->addLayout(btnRow);

    // ── 解析结果 ────────────────────────────
    auto *resultGroup = new QGroupBox("解析结果");
    auto *resultLayout = new QVBoxLayout(resultGroup);
    m_resultText = new QTextEdit;
    m_resultText->setReadOnly(true);
    m_resultText->setMaximumHeight(180);
    resultLayout->addWidget(m_resultText);
    layout->addWidget(resultGroup);

    layout->addStretch();
}

void FileSelectPage::onParseClicked() {
    QVariantMap paths;
    paths["esi_path"]      = m_esiEdit->text().trimmed().isEmpty() ? QVariant() : m_esiEdit->text().trimmed();
    paths["sdo_path"]      = m_sdoEdit->text().trimmed().isEmpty() ? QVariant() : m_sdoEdit->text().trimmed();
    paths["template_path"] = m_tplEdit->text().trimmed().isEmpty() ? QVariant() : m_tplEdit->text().trimmed();
    paths["output_dir"]    = m_outEdit->text().trimmed().isEmpty() ? "out" : m_outEdit->text().trimmed();

    if (paths["esi_path"].toString().isEmpty() && paths["sdo_path"].toString().isEmpty()) {
        m_resultText->setPlainText("错误: 请至少提供一个 ESI 或 SDO 文件。");
        return;
    }
    emit parseRequested(paths);
}

void FileSelectPage::showParseResult(const QVariantMap &info) {
    QStringList lines;
    for (auto it = info.begin(); it != info.end(); ++it)
        lines << QString("%1: %2").arg(it.key(), it.value().toString());
    m_resultText->setPlainText(lines.join("\n"));
}

QVariantMap FileSelectPage::getPaths() const {
    QVariantMap p;
    p["esi_path"]      = m_esiEdit->text().trimmed().isEmpty() ? QVariant() : m_esiEdit->text().trimmed();
    p["sdo_path"]      = m_sdoEdit->text().trimmed().isEmpty() ? QVariant() : m_sdoEdit->text().trimmed();
    p["template_path"] = m_tplEdit->text().trimmed().isEmpty() ? QVariant() : m_tplEdit->text().trimmed();
    p["output_dir"]    = m_outEdit->text().trimmed().isEmpty() ? "out" : m_outEdit->text().trimmed();
    return p;
}

void FileSelectPage::browseFile(QLineEdit *edit, const QString &filter) {
    QString path = QFileDialog::getOpenFileName(this, "选择文件", "", filter);
    if (!path.isEmpty()) edit->setText(path);
}

void FileSelectPage::browseDir(QLineEdit *edit) {
    QString path = QFileDialog::getExistingDirectory(this, "选择输出目录");
    if (!path.isEmpty()) edit->setText(path);
}
